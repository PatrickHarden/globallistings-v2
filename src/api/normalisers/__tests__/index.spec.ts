import residentialResponse from '../../../tests/fixtures/residential-response.json';
import commercialResponse from '../../../tests/fixtures/commercial-response.json';
import { normaliseApiResponse } from '../index';
import {
  withHighlights,
  highlights,
  propertiesToApiResponse,
  withStrapline,
  straplines,
  longDescriptions,
  withLongDescription,
  emptyProperty,
  withAddress,
  address,
  withLocation,
  pointLocation,
} from '../../../tests/factories/propertyFactory';
import { path } from 'ramda';

// Silence console error so it doesn't flood the console
// TODO remove this once we have the strategy to push those to a error logging system
beforeEach(() => {
  console.error = jest.fn();
});

describe('map propertylistings api response', () => {
  it('should map correctly for residential data', () => {
    const data = normaliseApiResponse('en-GB')(residentialResponse);
    expect(data).toMatchSnapshot();
    expect(data.result).toEqual(['GB-ReapIT-cbrrps-CAL180002']);
  });

  it('should map correctly for commercial data', () => {
    const data = normaliseApiResponse('en-GB')(commercialResponse as any);
    expect(data).toMatchSnapshot();
  });

  const baseProperty = {
    ...withAddress(address),
    ...withLocation(pointLocation),
  };

  const MOCK_ID = 'MOCK_ID';
  const toData = (data: any) =>
    propertiesToApiResponse([{ ...baseProperty, ...data, 'Common.PrimaryKey': MOCK_ID }]);

  describe('with multilingual property:', () => {
    const firstPropertyPath = ['entities', 'properties', MOCK_ID];
    const propPath = (p: any) => path(firstPropertyPath.concat(p));

    [
      // highlights
      {
        data: toData(withHighlights([highlights.gym.en])),
        locale: 'en-GB',
        path: propPath(['highlights']),
        expected: [highlights.gym.en['Common.Text']],
        desc: 'english highlights for en-GB locale',
        group: 'highlights',
      },
      {
        data: toData(withHighlights([highlights.gym.fr])),
        locale: 'fr-FR',
        path: propPath(['highlights']),
        expected: [highlights.gym.fr['Common.Text']],
        desc: 'french highlights for fr-FR locale',
        group: 'highlights',
      },
      {
        data: toData(withHighlights([highlights.gym.en, highlights.gym.fr])),
        locale: 'fr-FR',
        path: propPath(['highlights']),
        expected: [highlights.gym.fr['Common.Text']],
        desc: 'french highlights for fr-FR locale with multilingual highlights',
        group: 'highlights',
      },
      {
        data: toData(withHighlights([highlights.gym.fr, highlights.gym.en])),
        locale: 'fr-FR',
        path: propPath(['highlights']),
        expected: [highlights.gym.fr['Common.Text']],
        desc: 'french highlights for fr-FR locale with multilingual highlights switched around',
        group: 'highlights',
      },

      // strapline
      {
        data: toData(withStrapline([straplines.en])),
        locale: 'en-GB',
        path: propPath(['description', 'strapline']),
        expected: straplines.en['Common.Text'],
        desc: 'english strapline for en-GB locale',
        group: 'strapline',
      },
      {
        data: toData(withStrapline([straplines.fr])),
        locale: 'fr-FR',
        path: propPath(['description', 'strapline']),
        expected: straplines.fr['Common.Text'],
        desc: 'french strapline for fr-FR locale',
        group: 'strapline',
      },
      {
        data: toData(withStrapline([straplines.fr, straplines.en])),
        locale: 'fr-FR',
        path: propPath(['description', 'strapline']),
        expected: straplines.fr['Common.Text'],
        desc: 'french strapline for fr-FR locale with multilingual highlights',
        group: 'strapline',
      },
      {
        data: toData(withStrapline([straplines.en, straplines.fr])),
        locale: 'fr-FR',
        path: propPath(['description', 'strapline']),
        expected: straplines.fr['Common.Text'],
        desc: 'french strapline for fr-FR locale with multilingual highlights switched around',
        group: 'strapline',
      },

      // long description
      {
        data: toData(withLongDescription([longDescriptions.en])),
        locale: 'en-GB',
        path: propPath(['description', 'longDescription']),
        expected: longDescriptions.en['Common.Text'],
        desc: 'english longDescription for en-GB locale',
        group: 'longDescription',
      },
      {
        data: toData(withLongDescription([longDescriptions.fr])),
        locale: 'fr-FR',
        path: propPath(['description', 'longDescription']),
        expected: longDescriptions.fr['Common.Text'],
        desc: 'french longDescription for fr-FR locale',
        group: 'longDescription',
      },
      {
        data: toData(withLongDescription([longDescriptions.fr, longDescriptions.en])),
        locale: 'fr-FR',
        path: propPath(['description', 'longDescription']),
        expected: longDescriptions.fr['Common.Text'],
        desc: 'french longDescription for fr-FR locale with multilingual highlights',
        group: 'longDescription',
      },
      {
        data: toData(withLongDescription([longDescriptions.en, longDescriptions.fr])),
        locale: 'fr-FR',
        path: propPath(['description', 'longDescription']),
        expected: longDescriptions.fr['Common.Text'],
        desc:
          'french longDescription for fr-FR locale with multilingual highlights switched around',
        group: 'longDescription',
      },
    ].forEach(td => {
      describe(td.group, () => {
        it(`should return ${td.desc}`, () => {
          const result = normaliseApiResponse(td.locale)(td.data as any);
          expect(td.path(result)).toEqual(td.expected);
        });
      });
    });
  });

  describe('with missing data', () => {
    it('should handle gracefully', () => {
      expect(() => normaliseApiResponse('en-GB')(emptyProperty as any)).not.toThrow();
    });
  });

  /**
   * TODO write some tests to check that we log info about missing properties
   * This can be a bit tricky as dependant on the property some fields can become mandatory or not.
   * Will need to wait until we have more details about what is mandatory and what is not
   */
});
