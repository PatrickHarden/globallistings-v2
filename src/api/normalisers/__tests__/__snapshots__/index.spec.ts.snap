// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map propertylistings api response should map correctly for commercial data 1`] = `
Object {
  "entities": Object {
    "properties": Object {
      "GB-Plus-467051": Object {
        "address": Object {
          "Country": "UK",
          "Line1": "India House",
          "Line2": "45 Curlew Street",
          "Locality": "London",
          "PostCode": "SE1 2ND",
        },
        "charges": Array [
          Object {
            "amount": 15,
            "amountKind": "Value",
            "chargeKind": "BusinessRates",
            "chargeModifier": undefined,
            "currencyCode": "GBP",
            "interval": "Annually",
            "onApplication": undefined,
            "perUnit": "sqft",
            "taxModifier": "None",
            "year": 2019,
          },
          Object {
            "amount": 9.22,
            "amountKind": "Value",
            "chargeKind": "ServiceCharge",
            "chargeModifier": undefined,
            "currencyCode": "GBP",
            "interval": "Annually",
            "onApplication": undefined,
            "perUnit": "sqft",
            "taxModifier": "None",
            "year": 2018,
          },
          Object {
            "amount": undefined,
            "amountKind": undefined,
            "chargeKind": "Rent",
            "chargeModifier": "OnApplication",
            "currencyCode": undefined,
            "interval": undefined,
            "onApplication": true,
            "perUnit": undefined,
            "taxModifier": undefined,
            "year": undefined,
          },
        ],
        "contacts": Array [
          Object {
            "email": "+44 207 182 2577",
            "name": "Ben Deacon",
            "phone": "Ben.Deacon@cbre.com",
          },
          Object {
            "email": "+44 207 182 2165",
            "name": "Alex Kerr",
            "phone": "alex.kerr@cbre.com",
          },
        ],
        "description": Object {
          "longDescription": undefined,
          "strapline": undefined,
        },
        "highlights": Array [
          "Perimeter Comfort Cooling",
          "Raised Floors",
          "LED Lighting",
          "Commissionaire",
        ],
        "id": "GB-Plus-467051",
        "info": Object {
          "energyPerformance": undefined,
          "floorPlans": undefined,
        },
        "location": Object {
          "lat": 51.50250459999999,
          "lon": -0.07365300000003572,
        },
        "numberOfBedrooms": 0,
        "photos": Object {
          "hero": Object {
            "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
            "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
          },
          "main": undefined,
        },
        "size": Object {
          "max": Object {
            "amount": 6655,
            "units": "sqft",
          },
          "min": Object {
            "amount": 6655,
            "units": "sqft",
          },
        },
      },
    },
  },
  "result": Array [
    "GB-Plus-467051",
  ],
}
`;

exports[`map propertylistings api response should map correctly for residential data 1`] = `
Object {
  "entities": Object {
    "properties": Object {
      "GB-ReapIT-cbrrps-CAL180002": Object {
        "address": Object {
          "Country": "GB",
          "Line1": "Ariel House",
          "Line2": "144 Vaughan Way",
          "Locality": "London Dock",
          "PostCode": "E1W",
        },
        "charges": Array [
          Object {
            "amount": 595,
            "amountKind": undefined,
            "chargeKind": "Rent",
            "chargeModifier": undefined,
            "currencyCode": "GBP",
            "interval": "Weekly",
            "onApplication": undefined,
            "perUnit": "Whole",
            "taxModifier": "None",
            "year": undefined,
          },
        ],
        "contacts": Array [
          Object {
            "email": "+44 (0) 20 7205 2201",
            "name": "Canary Wharf Lettings",
            "phone": "canarywharf.lettings@cbre.com",
          },
        ],
        "description": Object {
          "longDescription": "The long awaited London Dock development has finally arrived and you can at last secure your place in it. 

Situated in the heart of the capital, just moments from the extremely desirable St Katharine Docks, this exciting new destination offers apartments and penthouses with the finest level of residents' facilities, a private suite of hotel style amenities including state-of- the-art gymnasium, swimming pool, squash court, virtual golf, treatment room, sauna, steam room, private screening room and residents lounge complimented by 24 hour concierge. 

The brand new third floor one bedroom flat is finished to an immaculate standard. The open plan kitchen and living room is a great size and has attractive wooden flooring as well as all the modern conveniences. The floor to ceiling window allows plenty of light into the flat and access to a huge private balcony which overlooks the square. There is also a wall mounted flat screen television.

The bedroom has a built in double wardrobe offering plenty of storage and meaning the ample space available is not compromised.

The bathroom is a joy and has underfloor heating and there is storage in the hallway which houses the washer dryers and flat controls meaning that they do not hinder the aesthetic or practical enjoyment of the property.

Surrounded by  beautifully landscaped public spaces, shops, bars and restaurants. Tower Hill underground is a short walk away meaning that there is easy access to the District and Circle Line. This makes trips to Chelsea, Victoria and Westminster a short affair. a Tower Hill Gateway DLR is also within walking distance so London Dock offers quick and easy access to the main financial centres of the City and Canary Wharf, as well as the exciting cultural and creative neighbourhoods of Shoreditch, Whitechapel and Bermondsey. The location will work fantastically for students and professionals.

For more information about our fees please visit our web-site's 'Tenant Fees' page located under the Rent tab.",
          "strapline": "An amazing one bed to rent in London&#039;s newest exclusive development, E1W",
        },
        "highlights": Array [
          "Golf Room",
          "Balcony",
          "Gym",
          "Furnished",
          "Swimming Pool",
          "24 hour Concierge",
        ],
        "id": "GB-ReapIT-cbrrps-CAL180002",
        "info": Object {
          "energyPerformance": Array [
            "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/6db2de5d/CAL180002_05.png",
          ],
          "floorPlans": Array [
            "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/39549074/CAL180002_22.jpg",
          ],
        },
        "location": Object {
          "lat": 51.505794,
          "lon": -0.060134,
        },
        "numberOfBedrooms": 1,
        "photos": Object {
          "hero": Object {
            "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/a860e9a6/CAL180002_06_Photo_1_large.jpg",
            "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/a860e9a6/CAL180002_06_Photo_1_small.jpg",
          },
          "main": Array [
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/a860e9a6/CAL180002_06_Photo_1_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/a860e9a6/CAL180002_06_Photo_1_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cdc6e57e/CAL180002_07_Photo_2_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cdc6e57e/CAL180002_07_Photo_2_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/c48f7da7/CAL180002_09_Photo_3_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/c48f7da7/CAL180002_09_Photo_3_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/1bc45f97/CAL180002_08_Photo_4_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/1bc45f97/CAL180002_08_Photo_4_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/342bd9b2/CAL180002_10_Photo_5_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/342bd9b2/CAL180002_10_Photo_5_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/be916afa/CAL180002_11_Photo_6_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/be916afa/CAL180002_11_Photo_6_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/ddeac696/CAL180002_12_Photo_7_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/ddeac696/CAL180002_12_Photo_7_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cbb349dc/CAL180002_13_Photo_8_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cbb349dc/CAL180002_13_Photo_8_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/500cd96d/CAL180002_14_Photo_9_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/500cd96d/CAL180002_14_Photo_9_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/e9e1ed42/CAL180002_15_Photo_10_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/e9e1ed42/CAL180002_15_Photo_10_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/83d0dae7/CAL180002_16_Photo_11_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/83d0dae7/CAL180002_16_Photo_11_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/32a4fc4b/CAL180002_17_Photo_12_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/32a4fc4b/CAL180002_17_Photo_12_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/4ac16bc3/CAL180002_18_Photo_13_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/4ac16bc3/CAL180002_18_Photo_13_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cff83809/CAL180002_19_Photo_14_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/cff83809/CAL180002_19_Photo_14_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/e96bcaca/CAL180002_20_Photo_15_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/e96bcaca/CAL180002_20_Photo_15_small.jpg",
            },
            Object {
              "large": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/bfff6f9a/CAL180002_21_Photo_16_large.jpg",
              "small": "/resources/fileassets/GB-ReapIT-cbrrps-CAL180002/bfff6f9a/CAL180002_21_Photo_16_small.jpg",
            },
          ],
        },
      },
    },
  },
  "result": Array [
    "GB-ReapIT-cbrrps-CAL180002",
  ],
}
`;
